jormungandr/src/rest/server/mod.rs

```diff
@@ -10,40 +10,88 @@ use actix_web::{
     actix::{Addr, System},
     server::{self, IntoHttpHandler, StopServer},
 };
+use futures::sync::oneshot::{self, Receiver};
+use futures::{Async, Future, Poll};
 use native_tls::{Identity, TlsAcceptor};
 use std::{
     fs,
     net::{SocketAddr, ToSocketAddrs},
     path::PathBuf,
+    sync::mpsc,
+    thread,
 };
 
 pub type ServerResult<T> = Result<T, Error>;
 
-#[derive(Clone)]
+/// A future that resolves when server shuts down
+/// Dropping Server causes its shutdown
 pub struct Server {
+    stopper: ServerStopper,
+    stop_receiver: Receiver<()>,
+}
+
+#[derive(Clone)]
+pub struct ServerStopper {
     addr: Addr<ActixServer>,
 }
 
 impl Server {
-    pub fn run<F, H>(
+    pub fn start<F, H>(
         pkcs12: Option<PathBuf>,
         address: SocketAddr,
         handler: F,
-        server_receiver: impl FnOnce(Server),
-    ) -> ServerResult<()>
+    ) -> ServerResult<Server>
     where
         F: Fn() -> H + Clone + Send + 'static,
         H: IntoHttpHandler + 'static,
     {
         let tls = load_tls_acceptor(pkcs12)?;
-        let actix_system = System::builder().build();
-        let addr = start_server_curr_actix_system(address, tls, handler)?;
-        let server = Server { addr };
-        server_receiver(server);
-        actix_system.run();
-        Ok(())
+        let (server_sender, server_receiver) = mpsc::sync_channel::<ServerResult<Server>>(0);
+        thread::spawn(move || {
+            let actix_system = System::builder().build();
+            let (stop_sender, stop_receiver) = oneshot::channel();
+            let server_res =
+                start_server_curr_actix_system(address, tls, handler).map(move |addr| Server {
+                    stopper: ServerStopper { addr },
+                    stop_receiver,
+                });
+            let run_system = server_res.is_ok();
+            let _ = server_sender.send(server_res);
+            if run_system {
+                actix_system.run();
+            };
+            let _ = stop_sender.send(());
+        });
+        server_receiver
+            .recv()
+            .expect("Actix thread terminated before sending server handle")
+    }
+
+    pub fn stopper(&self) -> ServerStopper {
+        self.stopper.clone()
+    }
+}
+
+impl Future for Server {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self.stop_receiver.poll() {
+            Err(_) => Ok(Async::Ready(())),
+            Ok(ok) => Ok(ok),
+        }
     }
+}
+
+impl Drop for Server {
+    fn drop(&mut self) {
+        self.stopper.stop();
+        let _ = self.wait();
```

Comment:
not sure about that though...

Reply:
If there is a deadlock in there this is risky.

Reply:
Me neither. I've decided that it's better than having REST server randomly killed in the middle of network operation. Timeout would be the perfect solution, but at that point Tokio runtime is already down and it's not easy without its reactor.

Reply:
Wait, what? Why do you need to do this? It should be OK to drop "disgracefully", that's how cancellation of everything works in futures.

Reply:
If there is a graceful shutdown, it should be done with sending a `Shutdown` message to the task.

Reply:
:man_facepalming: that's so obvious!